torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +588,167 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
+    const AllToAllOptions& opts) {
+  checkSingleTensor(outputTensors);
+  checkSingleTensor(inputTensors);
+
+  if(outputSplitSizes.size() == 0 && inputSplitSizes.size() == 0) {
+    // We can use alltoall
+    if ((outputTensors[0].numel() != inputTensors[0].numel()) ||
+        (outputTensors[0].type() != inputTensors[0].type())) {
+      throw std::runtime_error("Tensors are not equal in size or data type");
+    }
+    checkSingleTensorHelper(inputTensors[0]);
+    checkSingleTensorHelper(outputTensors[0]);
+    if(outputTensors[0].size(0) % size_ != 0) {
+      throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+    }
+    
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoall(
+              srcdata.data_ptr(),
+              srcdata.numel()/size_,
+              mpiDatatype.at(srcdata.scalar_type()),
+              dstdata.data_ptr(),
+              dstdata.numel()/size_,
+              mpiDatatype.at(dstdata.scalar_type()),
+              pgComm_));
+      };
+    auto entry = std::unique_ptr<WorkEntry>(
+        new WorkEntry(&inputTensors, &outputTensors, std::move(runFunc)));
+    return enqueue(std::move(entry));
+  } else {
+    // Need alltoallv
+    auto checkSplitSizes = [this](std::vector<int> split_sizes, at::Tensor &tensor) {
+      if(split_sizes.size() == 0) {
+        if(tensor.size(0) % size_ != 0) {
+          throw std::runtime_error("Tensor's dim 0 does not divide equally across group size");
+        }
+      } else {
+        if(split_sizes.size() != size_) {
+          throw std::runtime_error("Number of tensor splits not equal to group size");
+        }
+        int sum = std::accumulate(split_sizes.begin(), split_sizes.end(), 0);
+        if(sum != tensor.size(0)) {
+          throw std::runtime_error("Split sizes doesn't match total dim 0 size");
+        }
+      }
+    };
+    checkSplitSizes(inputSplitSizes, inputTensors[0]);
+    checkSplitSizes(outputSplitSizes, outputTensors[0]);
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this, inputSplitSizes, outputSplitSizes](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        std::vector<int> send_lengths(size_);
+        std::vector<int> recv_lengths(size_);
+        std::vector<int> send_offsets(size_);
+        std::vector<int> recv_offsets(size_);
+        auto computeLengthsAndOffsets = [this](std::vector<int> split_sizes, at::Tensor &tensor, std::vector<int> &lengths, std::vector<int> &offsets) {
+          if(split_sizes.size() == 0) {
+            int split_size = tensor.numel() / size_;
```

Comment:
This can overflow for large tensors and MPI will complain about negative splits and crash the program. [unfortunately looks like the standard still specifies displacements to be 32-bit ints for backwards compatibility]. To be 100% safe, I suggest running a sanity check collective right before calling MPI to make sure all arguments are within range. You can also use something like the `SafeInt` library to avoid writing a lot of these checks explicitly (https://github.com/dcleblanc/SafeInt/blob/master/SafeInt.hpp)

Reply:
Yes, agree. Will change `std::vector<int>&` to `std::vector<int64_t>&` in the arguments and add check for overflow.